File: web/tomato/dynimg.py
# -*- coding: utf-8 -*-

# ToMaTo (Topology management software) 
# Copyright (C) 2010 Dennis Schwerdel, University of Kaiserslautern
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>

import os
from django.shortcuts import redirect

def dynimg(request,size,objtype,arg1,arg2):
	"""
	Uses 2 arguments to build a path to an image file for a specific object type. It then checks whether this image file exists. On success, it redirects to this file. If it does not exist, it redirects to a generic image file for this object type.
	for templates: tech/size/subtype/name
	for networks:  network/size/kind/none
	for vpns:	  vpn/size/kind/none
	"""
	filepath_pref = ""
	url_alt = ""
	
	
	# handle devices
	if objtype=="openvz" or objtype=="kvmqm" or objtype=="repy":
		subtype=arg1
		template_name=arg2
		
		#handle non-(repy-network) devices
		if objtype!="repy" or subtype=="device":
			found_templ = False
			for i in ['floodlight']:
				if i in template_name:
					template_name = i
					found_templ = True
			#if the template is known, first try to find a template-specific image. On error, redirect to a non-template dynimg url
			if found_templ:
				filepath_pref="img/"+objtype+"_"+template_name+size+".png"
				url_alt = "/dynimg/"+objtype+"/"+size+"/"+subtype+"/none"
			#if the template is not known, try to match the subtype
			else:
				filepath_pref="img/"+objtype+"_"+subtype+size+".png"
				url_alt ="/img/"+objtype+size+".png"
		# handle repy network devices
		else:
			filepath_pref="img/switch_repy32.png"
			url_alt="/img/network32.png"
		
	# handle networks
	elif objtype=="network" or objtype=="vpn":
		kind=arg1
		for i in ["openflow"]:
			if i in kind:
				kind = i
		filepath_pref="img/"+kind+size+".png"
		url_alt="/img/"+objtype+size+".png"
		
	# handle everything else
	else:
		filepath_pref="img/"+objtype+size+".png"
		url_alt="/img/"+objtype+".png"
		
		
	# now, check whether the file exists and redirect to the preferred or alternate url
	url_pref="/"+filepath_pref
	prefix = os.path.dirname(os.path.realpath(__file__))
	if os.path.exists(os.path.join(prefix, filepath_pref)):
		return redirect(url_pref)
	else:
		return redirect(url_alt)
		
	

File: web/tomato/js/dynimg.js
//does a blocking ajax call to get the contents of the img directory
function dynimg_getlsfromajax() {
	var xmlhttp = new XMLHttpRequest();
	xmlhttp.open("GET","/img_ls",false);
	xmlhttp.send();
	return JSON.parse(xmlhttp.responseText);
}

var dynimg_ls = null;
//proxy pattern. get the contents of the img directory, and keep it in memory.
function dynimg_getls() {
	if (dynimg_ls==null) {
		dynimg_ls=dynimg_getlsfromajax();
	}
	return dynimg_ls;
}

//check whether a file with a certain name exists inside the server's img directory
function dynimg_imgExists(filename) {
	ls=dynimg_getls();
	for(var i=0; i<ls.length; i++) {
		if (ls[i] == filename) {
			return true;
		}
	}
	return false;
}

//get the url for an image
function dynimg(size,objtype,arg1,arg2) {
	/*
	Uses 2 arguments to build a path to an image file for a specific object type. It then checks whether this image file exists. On success, it redirects to this file. If it does not exist, it redirects to a generic image file for this object type.
	for templates: objtype==openvz|kvmqm|repy  arg1:subtype  arg2:templatename|null  
	for networks:  objtype==network  arg1:kind  arg2:null   
	for vpns:	   objtype==vpn  arg1:kind  arg2:null
	*/

	//templates
	if (objtype=="openvz" || objtype=="kvmqm" || objtype=="repy") {
		var subtype=arg1;
		var template_name=arg2;
		
		//non-(repy-network) templates
		if (objtype!="repy" || subtype=="device") {
			if (template_name!=null) {
				var filename=objtype+"_"+template_name+size+".png";
				if (dynimg_imgExists(filename)) {
					return "/img/"+filename;
				}
			}
			//if no file exists for the template, or it hasn't been found, look for a generic file for the subtype.
			var filename = objtype+"_"+subtype+size+".png";
			if (dynimg_imgExists(filename)) {
				return "/img/"+filename;
			}
			//if this also didn't work, return a file for the tech
			return "/img/"+subtype+size+".png";			
		}
		
		//repy-network templates
		else {
			var filename="switch_repy"+size+".png";
			if (dynimg_imgExists(filename)) {
				return "/img/"+filename;
			}
			return "/img/network"+32+".png";
		}
	}
	

	
	
	//networks or vpns
	if (objtype=="network" || objtype=="vpn") {
		var kind=arg1;
		var filename=kind+size+".png"
		if (dynimg_imgExists(filename)) {
			return "/img/"+filename;
		}
		var filename = objtype+size+".png";
		if (dynimg_imgExists(filename)) {
			return "/img/"+filename;
		}
		return "/img/network"+size+".png";
	}
	
	return null;
}
